### PR TITLE
Capi machines

### PR DESCRIPTION
**Pull request by enxebre**
This drops openshift machine API in favour of capi machine management.
Move guestInfraCluster to externalInfraCluster to satisfy kubernetes-sigs/cluster-api-provider-aws#2124
Run capa controller manager
Enable automation for creating a secret with aws credentials for the guest cluster/machine controller.

### CODE REVIEW

File: hypershift-operator/controllers/externalinfracluster_controller.go [hunk 1]
```diff
@@ -55,24 +55,24 @@ func (r *GuestClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
-func (r *GuestClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *ExternalInfraClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	r.Log = ctrl.LoggerFrom(ctx)
 	r.Log.Info("Reconciling")
 
-	// Fetch the GuestCluster instance
-	guestCluster := &hyperv1.GuestCluster{}
-	err := r.Client.Get(ctx, req.NamespacedName, guestCluster)
+	// Fetch the ExternalInfraCluster instance
+	ExternalInfraCluster := &hyperv1.ExternalInfraCluster{}
```
Comment (csrwng) [inline hypershift-operator/controllers/externalinfracluster_controller.go]: should use lowercase for local variable

File: hypershift-operator/controllers/nodepool_controller.go [hunk 2]
```diff
@@ -161,53 +153,112 @@ func GetOClusterByName(ctx context.Context, c client.Client, namespace, name str
...
+						SecureSecretsBackend:       "secrets-manager",
+					},
+					IAMInstanceProfile: instanceProfile,
+					InstanceType:       instanceType,
+					AMI: capiaws.AWSResourceReference{
+						ID: k8sutilspointer.StringPtr(AMI),
+					},
+					Subnet: &capiaws.AWSResourceReference{
+						Filters: []capiaws.Filter{
+							capiaws.Filter{
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool_controller.go]: redundant `capi.Filter`... `{` is enough here

File: hypershift-operator/controllers/nodepool_controller.go [hunk 3]
```diff
@@ -161,53 +153,112 @@ func GetOClusterByName(ctx context.Context, c client.Client, namespace, name str
...
+			Name:      resourcesName,
+			Namespace: nodePool.GetNamespace(),
+		},
+		Spec: capiaws.AWSMachineTemplateSpec{
+			Template: capiaws.AWSMachineTemplateResource{
+				Spec: capiaws.AWSMachineSpec{
+					UncompressedUserData: k8sutilspointer.BoolPtr(true),
+					CloudInit: capiaws.CloudInit{
+						InsecureSkipSecretsManager: true,
+						SecureSecretsBackend:       "secrets-manager",
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool_controller.go:198]: Just curious, what does this do?

File: hypershift-operator/controllers/hosted_controlplane_controller.go [hunk 4]
```diff
@@ -168,25 +168,36 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+			Namespace: req.Namespace,
+			Name:      fmt.Sprintf("%s-user-data", req.Name),
 		},
 	}
 	if err := waitForDeletion(ctx, r.Log, r.Client, machineSetConfig); err != nil && !apierrors.IsNotFound(err) {
 		return fmt.Errorf("failed to delete machineset secret %s: %w", machineSetConfig.Name, err)
 	}
 	r.Log.Info("deleted machineset secret", "name", machineSetConfig.Name)
 
+	kubeconfig := &corev1.Secret{
```
Comment (csrwng) [inline hypershift-operator/controllers/hosted_controlplane_controller.go]: should we just set the owner reference on the secret and then just let the delete cascade?

File: hypershift-operator/controllers/externalinfracluster_controller.go [hunk 5]
```diff
@@ -162,13 +162,13 @@ func (r *GuestClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
```
Comment (csrwng) [inline hypershift-operator/controllers/externalinfracluster_controller.go:159]: does the comment above this line still apply?

File: config/example-cluster/cluster.yaml [hunk 6]
```diff
@@ -5,7 +5,7 @@ metadata:
 spec:
   release:
     image: quay.io/openshift-release-dev/ocp-release:4.6.7-x86_64
-  initialComputeReplicas: 2
+  initialComputeReplicas: 0
```
Comment (csrwng) [inline config/example-cluster/cluster.yaml]: is there a reason you're changing this?

File: config/example-cluster/kustomization.yaml [hunk 7]
```diff
@@ -18,5 +18,11 @@ secretGenerator:
   options:
     disableNameSuffixHash: true
   files:
-  - id_rsa.pub=ssh-key
+    - id_rsa.pub=ssh-key
```
Comment (csrwng) [inline config/example-cluster/kustomization.yaml]: is there a reason for the extra indent?

File: README.md [hunk 8]
```diff
@@ -27,7 +27,8 @@ $ make uninstall
 First, create the following files containing secrets used by the example cluster:
 
 - `config/example-cluster/pull-secret` a valid pull secret for image pulls.
-- `config/example-cluster/ssh-key` an SSH public key for guest node access
+- `config/example-cluster/ssh-key` an SSH public key for guest node access.
+- `config/example-cluster/aws-creds` a base64 encoded [aws credentials file](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html).
```
Comment (csrwng) [inline README.md]: I don't think this is the right place for this secret. It's a secret required by the capa controller deployment, but these secrets are created along with an example cluster, so you won't get a running capa controller until you create the example cluster.

Comment (csrwng) [reply README.md]: Also, I don't think it needs to be base64 encoded. Kustomize should base64 encode it.

**Review by csrwng — COMMENTED** [review]
@enxebre looks awesome. Some comments.

File: README.md (hunk 8 above)
Comment (enxebre) [reply README.md]: I moved the capa controller down to the operator control. Also the secret is copied now into the target ns similar to what we do with the pull-secret.

File: config/hypershift-operator/release-info.json [hunk 9]
```diff
@@ -0,0 +1,2158 @@
+{
```
Comment (csrwng) [inline config/hypershift-operator/release-info.json]: This file is not needed anymore

File: hypershift-operator/controllers/controlplane.go [hunk 10]
```diff
@@ -67,6 +69,17 @@ func (r *HostedControlPlaneReconciler) ensureControlPlane(ctx context.Context, h
...
 	}
+
+	var providerCredsSecret corev1.Secret
+	err = r.Client.Get(ctx, client.ObjectKey{Namespace: hcp.Namespace, Name: hcp.Spec.ProviderCreds.Name}, &providerCredsSecret)
+	if err != nil {
+		return fmt.Errorf("failed to get provider creds %s: %w", hcp.Spec.ProviderCreds.Name, err)
+	}
+	providerCredsData, hasProviderCredsData := providerCredsSecret.Data["credentials"]
+	if !hasProviderCredsData {
+		return fmt.Errorf("pull secret %s is missing the .dockerconfigjson key", hcp.Spec.PullSecret.Name)
```
Comment (csrwng) [inline hypershift-operator/controllers/controlplane.go]: wrong error message

File: hypershift-operator/controllers/nodepool_controller.go [hunk 11]
```diff
@@ -5,21 +5,20 @@ import (
 	"fmt"
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool_controller.go:5]: The event recorder in line 56 should be for "nodepool-controller" https://github.com/openshift-hive/hypershift/pull/31/files#diff-804e9a40d4fe9658dc8326383ec33b1cc577ab536c2b8b6ec58fcbff122a1718L56

File: hypershift-operator/controllers/externalinfracluster_controller.go [hunk 12]
```diff
@@ -24,18 +24,18 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-type GuestClusterReconciler struct {
+type ExternalInfraClusterReconciler struct {
```
Comment (csrwng) [inline hypershift-operator/controllers/externalinfracluster_controller.go:27]: The event recorder in line 53 should be for "external-infra-cluster-controller" ... https://github.com/openshift-hive/hypershift/pull/31/files#diff-2234353920a980ab201b2fc06e4ef30d221189c9655e2f8922e5877cff1b5325L53

**Review by csrwng — COMMENTED** [review]
@enxebre looking good... gave it another pass. Just thinking about the name ExternalInfraCluster ... should we not name it something more specific to AWS ? like 'HostedAWSInfra' or something like that? 
Also, since you already made the change upstream to decouple things, and we get the node linking for free with this, I don't see a reason not to merge this when ready. @ironcladlou do you agree?

**Comment by derekwaynecarr** [comment]
just capturing follow-up from our discussion this morning:

- do we use the version field on machine spec/status?
- is decoupling for providers other than aws in flight?

**Comment by enxebre** [comment]
>do we use the version field on machine spec/status?

The version field is purely an option to override the bootstrap provider value if any https://github.com/kubernetes-sigs/cluster-api/blob/f5939fd5246e7f1682399e8004e5fd[…]65240/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

We don't use it atm, since we generate static ignition which is then injected as a secret into the machineSpec. If we ever have a bootstrap provider for this we might want to use it.

**Comment by ironcladlou** [comment]
@enxebre thanks for all this work!

**Pull request merged**